Should be possible to use GCS without authentication

My use-case is local development and testing using [fake-gcs-server](https://github.com/fsouza/fake-gcs-server). Auth is not required to use its GCS-compatible API.

However, Trino does not appear to have a way to not attempt authentication with Google's servers, even when setting the `gcs.endpoint` property.

If no auth-related properties (like `gcs.json-key`) are set, then [GcsStorageFactory](https://github.com/trinodb/trino/blob/475/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java#L99) calls `GoogleCredentials.getApplicationDefault()`.

However, this method throws an exception if no credentials are found:

```
Caused by: java.io.IOException: Your default credentials were not found. To set up Application Default Credentials for your environment, see https://cloud.google.com/docs/authentication/external/set-up-adc.
```

In contrast, if [`StorageOptions.Builder#setCredentials`](https://github.com/trinodb/trino/blob/475/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java#L116) is simply never called, the [underlying GCS library](https://github.com/googleapis/sdk-platform-java/blob/v2.53.0/java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java#L385-L391) populates it by calling `GoogleCredentials.getApplicationDefault()` but ignoring any exceptions it throws.

I have a temporary workaround to just not call `StorageOptions.Builder#setCredentials` in `GcsStorageFactory` if `gcs.endpoint` is set:
```
diff --git a/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java b/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
index f8ea12c452..b6b20537fd 100644
--- a/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
+++ b/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
@@ -93,6 +93,9 @@ public class GcsStorageFactory
                     credentials = GoogleCredentials.fromStream(inputStream).createScoped(DEFAULT_SCOPES);
                 }
             }
+            else if (endpoint.isPresent()) {
+                credentials = null;
+            }
             else {
                 credentials = jsonGoogleCredential.orElseGet(() -> {
                     try {
@@ -110,10 +113,13 @@ public class GcsStorageFactory

             endpoint.ifPresent(storageOptionsBuilder::setHost);

+            if (credentials != null) {
+                storageOptionsBuilder.setCredentials(credentials);
+            }
+
             // Note: without uniform strategy we cannot retry idempotent operations.
             // The trino-filesystem api does not violate the conditions for idempotency, see https://cloud.google.com/storage/docs/retry-strategy#java for details.
             return storageOptionsBuilder
-                    .setCredentials(credentials)
                     .setStorageRetryStrategy(getUniformStorageRetryStrategy())
                     .setRetrySettings(RetrySettings.newBuilder()
                             .setMaxAttempts(maxRetries + 1)
```

The three solutions I see are, in order of preference:
- where `GcsStorageFactory` [currently calls](https://github.com/trinodb/trino/blob/475/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java#L97-L104) `getApplicationDefault()`, just set `credentials` to null instead, and if it is indeed null, do not call `StorageOptions.Builder#setCredentials`, allowing the underlying library's behavior to apply
- suppress exceptions thrown by `GoogleCredentials.getApplicationDefault()` like the GCS client library itself does by default
- add an additional config property to disable auth for GCS altogether, perhaps resulting in passing `NoCredentials.getInstance()` as the credentials

I'm happy to open a PR for this.

--- a/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
+++ b/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
@@ -99,7 +99,8 @@ public class GcsStorageFactory
                         return GoogleCredentials.getApplicationDefault();
                     }
                     catch (IOException e) {
-                        throw new UncheckedIOException(e);
+                        // This is consistent with the GCP SDK when no credentials are available in the environment
+                        return null;
                     }
                 });
             }
@@ -108,12 +109,15 @@ public class GcsStorageFactory
                 storageOptionsBuilder.setProjectId(projectId);
             }
 
+            if (credentials != null) {
+                storageOptionsBuilder.setCredentials(credentials);
+            }
+
             endpoint.ifPresent(storageOptionsBuilder::setHost);
 
             // Note: without uniform strategy we cannot retry idempotent operations.
             // The trino-filesystem api does not violate the conditions for idempotency, see https://cloud.google.com/storage/docs/retry-strategy#java for details.
             return storageOptionsBuilder
-                    .setCredentials(credentials)
                     .setStorageRetryStrategy(getUniformStorageRetryStrategy())
                     .setRetrySettings(RetrySettings.newBuilder()
                             .setMaxAttempts(maxRetries + 1)

--- a/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsStorageFactory.java
+++ b/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsStorageFactory.java
@@ -1,0 +1,46 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.gcs;
+
+import com.google.auth.Credentials;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import io.trino.spi.security.ConnectorIdentity;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+final class TestGcsStorageFactory
+{
+    @Test
+    void testDefaultCredentials()
+            throws Exception
+    {
+        Credentials expectedCredentials = StorageOptions.newBuilder().build().getCredentials();
+
+        // No credentials options are set
+        GcsFileSystemConfig config = new GcsFileSystemConfig();
+
+        GcsStorageFactory storageFactory = new GcsStorageFactory(config);
+
+        Credentials actualCredentials;
+        try (Storage storage = storageFactory.create(ConnectorIdentity.ofUser("test"))) {
+            actualCredentials = storage.getOptions().getCredentials();
+        }
+
+        assertThat(actualCredentials)
+                .as("if credentials are not explicitly configured, should have same behavior as the GCS client")
+                .isEqualTo(expectedCredentials);
+    }
+}

